container-search/src/main/java/com/yahoo/search/dispatch/SearchPath.java

```diff
@@ -0,0 +1,253 @@
+ */
+public class SearchPath {
+    /**
+     * Parse the search path and select nodes from the given cluster based on it.
+     *
+     * @param searchPath
+     *            unparsed search path expression (see: model.searchPath in Search
+     *            API reference)
+     * @param cluster
+     *            the search cluster from which nodes are selected
+     * @throws InvalidSearchPathException
+     *             if the searchPath is malformed
+     * @return list of nodes chosen with the search path, or an empty list in which
+     *         case some other node selection logic should be used
+     */
+    public static List<SearchCluster.Node> selectNodes(String searchPath, SearchCluster cluster) {
+        Optional<SearchPath> sp = SearchPath.fromString(searchPath);
+        if (sp.isPresent()) {
+            return sp.get().mapToNodes(cluster);
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    public static Optional<SearchPath> fromString(String path) {
+        if (path == null || path.isEmpty()) {
+            return Optional.empty();
+        }
+        if (path.indexOf(';') >= 0) {
+            return Optional.empty(); // multi-level not supported at this time
+        }
+        try {
+            SearchPath sp = parseElement(path);
+            if (sp.isEmpty()) {
+                return Optional.empty();
+            } else {
+                return Optional.of(sp);
+            }
+        } catch (NumberFormatException | InvalidSearchPathException e) {
+            throw new InvalidSearchPathException("Invalid search path: " + path);
+        }
+    }
+
+    private final List<Part> parts;
+    private final Integer row;
+
+    private SearchPath(List<Part> parts, Integer row) {
+        this.parts = parts;
+        this.row = row;
+    }
+
+    private List<SearchCluster.Node> mapToNodes(SearchCluster cluster) {
+        if (cluster.groups().isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        SearchCluster.Group group = selectGroup(cluster);
+
+        if (parts.isEmpty()) {
+            return group.nodes();
+        }
+        Set<Integer> wanted = new HashSet<>();
+        int max = group.nodes().size();
+        for (Part part : parts) {
+            wanted.addAll(part.matches(max));
+        }
+        // ordering by distribution key might not be equal to ordering in services.xml
+        List<SearchCluster.Node> sortedByDistKey = new ArrayList<>(group.nodes());
+        sortedByDistKey.sort(Comparator.comparingInt(SearchCluster.Node::key));
+
+        List<SearchCluster.Node> ret = new ArrayList<>();
+        for (int idx : wanted) {
+            ret.add(sortedByDistKey.get(idx));
+        }
+        return ret;
+    }
+
+    private boolean isEmpty() {
+        return parts.isEmpty() && row == null;
+    }
+
+    private Group selectGroup(SearchCluster cluster) {
+        // ordering by group id might not be equal to ordering in services.xml
```

Comment:
How about we keep this sorted in SearchCluster instead?

Reply:
I initially thought that the use of searchPath is so limited that doing more processing in the parsing phase is preferable. At the same time the matrix of nodes in SearchCluster doesn't really take all that much space either and will take care of these issues, so I'll do that. (Applies to the node list ordering too)